### PR TITLE
MRTK Optimize Window tool to automate perf settings

### PR DESCRIPTION
**Pull request by Troy-Ferrell**
## Overview
This change introduces an editor window that will help developers analyze lacking performance settings in their application development. 

Tool can be launched via:
Mixed Reality Toolkit > Utilities > Optimize Window

Features:
- Project Optimizations:
    - SIngle Pass Instanced Rendering
    - Depth Buffer Sharing
- Scene Optimizations:
    - Disable Global Illumination
    - Analyze scene for num of lights and lights casting shadows
    - Analyze scene for 5 largest geometry (by polygon count)
- Shader Optimizations:
    - Analyze assetdatabase for all materials with Unity Standard shader
    - Convert materials to MRTK Standard Shader

Performance recommendations updated based on performance target developer desires (AR headsets, VR standalone, VR Tethered). Lots of area of improvement in this vein. 

Note: Editor changes only. No breaking change to MRTK infrastructure

MRTK_Optimize1

MRTK_Optimize2

MRTK_Optimize3

## Changes
- Fixes: #4298 , #3927, #4137 

## Verification
Tested "Optimize Project" to change project settings
Tested "Optimize Lighting" to change global illumination
Tested "Analyze scene"
Tested for null replacement shader
Tested for null material or missing material
Tested undo for material conversions
Tested MRTK build to VS app folder
Tested closing & re-opening window
Tested docking window in other areas

### CODE REVIEW

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeUtils.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    public class MixedRealityOptimizeUtils
+    {
+        public static void SetDepthBufferSharing(bool enableDepthBuffer)
+        {
+            PlayerSettings.VROculus.sharedDepthBuffer = enableDepthBuffer;
+
+#if UNITY_2019
+        PlayerSettings.VRWindowsMixedReality.depthBufferSharingEnabled = enableDepthBuffer;
+
+        PlayerSettings.VRWindowsMixedReality.depthBufferFormat = set16BitDepthBuffer ?
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeUtils.cs]: Getting an error here in 2019.1.1, `The name 'set16BitDepthBuffer' does not exist in the current context`

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+        private List<Material> discoveredMaterials = new List<Material>();
+
+        // Internal structure to easily search mesh polycounts in scene
+        private struct MeshNode
+        {
+            public int polycount;
+            public MeshFilter filter;
+        }
+
+        protected enum PerformanceTarget
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: Super dee duper nit: Not loving the underscores in the enum values. They look ugly.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+            window.Show();
+        }
+
+        public void StartUp()
+        {
+            FindShaders();
+
+            string assetPath = "StandardAssets/Textures";
+
+            if (logoLightTheme == null)
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: Tangent: we really need to move this sort of thing into an MRTK editor GUI utility.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+        {
+            Project = 0,
+            Scene,
+            Shader
+        };
+
+        private Vector2 windowScrollPosition = new Vector2();
+        private bool showProjectOptimizations = true;
+        private bool showSceneOptimizations = true;
+        private bool showShaderOptimizations = true;
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: Getting a warning in 2019.1.1,  `The field 'MixedRealityOptimizeWindow.showShaderOptimizations' is assigned but its value is never used`

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+                        OptimizeScene();
+                    }
+                });
+
+                BuildSection("Live Scene Analysis", null, () =>
+                {
+                    EditorGUILayout.BeginHorizontal();
+                    EditorGUILayout.LabelField(lastAnalyzedTime == null ? "Click analysis button for MRTK to scan your currently opened scene" : "Scanned " + GetRelativeTime(lastAnalyzedTime));
+
+                    if (GUILayout.Button("Analyze Scene", GUILayout.Width(200f)))
```
Comment (Railboy) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: Super nit: This button is over-sized compared to other similar buttons (eg View Lighting Settings)

**Comment by Railboy** [comment]
This is a great start. I like how it was able to handle deleting assets mid-evaluation.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs (hunk 3 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: I know :) I was going to do after all my PRs had merged and just do one pass of making things cleaner/more re-usable across components. But I can add the utility now and then just do a follow PR on other stuff to re-use that

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeUtils.cs (hunk 1 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeUtils.cs]: Ah damn, sorry, I didn't push last commit.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+        }
+
+        private void RenderShaderOptimizations()
+        {
+            GUILayout.BeginVertical("Box");
+            EditorGUILayout.LabelField(this.ToolbarTitles[(int)ToolbarSection.Shader], BoldLargeTitle);
+            using (new EditorGUI.IndentLevelScope())
+            {
+                EditorGUILayout.LabelField("The Unity standard shader is generally not performant or optimized for Mixed Reality development. The MRTK Standard shader can be a more performant option."
+                + "This tool allows developers to discover and automatically convert materials in their project to the replacement shader option below. It is recommended to utilize the MRTK Standard Shader."
```
Comment (Cameron-Micka) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: Super Nit: Missing space between '.' and This

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+
+        private void ConvertMaterial(Material asset)
+        {
+            if (asset != null && CanConvertMaterial(asset))
+            {
+                asset.shader = replacementShader;
+            }
+        }
+
+        private bool CanConvertMaterial(Material asset)
```
Comment (Cameron-Micka) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: It was eye opening to see how many materials in the MRTK are still using Unity's Standard shader, this tool is awesome! Sadly some of the materials found were imbedded in others assets (often FBXs) and cannot be altered (Unity makes it looks like you can alter them, be when you re-open the editor your changes are lost). Maybe we should filter out materials whose asset path does not end in ".mat"? Or, we could flag them somehow and say "Hey! Extract these materials and please change their shader."

Comment (Railboy) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]: >It was eye opening to see how many materials in the MRTK are still using Unity's Standard shader

Right? I couldn't believe it either.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+            Scene,
+            Shader
+        };
+
+        private Vector2 windowScrollPosition = new Vector2();
+
+        // Project Optimizations
+        private bool singlePassInstanced = true;
+        private bool enableDepthBufferSharing = true;
+        //private bool enable16BitDepthBuffer = true;
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityOptimizeWindow.cs]:
```suggestion

```

**Pull request merged**